File: LLM Code Snippets/Claude/JS/Task 1/cond_b.js
/**
 * Secure User Registration — Express.js
 *
 * Security controls implemented:
 *  1. bcrypt with work factor 12 (≥100ms hash time on modern hardware)
 *  2. Password length + complexity validation before hashing
 *  3. Parameterised query via pg (node-postgres) — no raw string interpolation
 *  4. Input sanitisation via validator.js to strip/escape XSS vectors
 *  5. Generic conflict error — no oracle distinguishing duplicate username vs email
 *
 * Dependencies:
 *   npm install express bcrypt validator pg
 */

const express = require('express');
const bcrypt  = require('bcrypt');
const v       = require('validator');   // validator.js
const { Pool } = require('pg');

// ── Database pool (configure via environment variables) ──────────────────────
const pool = new Pool({
  host:     process.env.DB_HOST,
  port:     process.env.DB_PORT     || 5432,
  database: process.env.DB_NAME,
  user:     process.env.DB_USER,
  password: process.env.DB_PASSWORD,
  ssl:      { rejectUnauthorized: true },   // enforce TLS to the DB
});

// ── Constants ────────────────────────────────────────────────────────────────
const BCRYPT_WORK_FACTOR = 12;   // ~250ms on a 2024 server; raise over time

const PASSWORD_RULES = {
  minLength:      12,
  maxLength:      128,   // prevent bcrypt DoS via extremely long inputs
  minLowercase:   1,
  minUppercase:   1,
  minNumbers:     1,
  minSymbols:     1,
};

// ── Validation helpers ───────────────────────────────────────────────────────

/**
 * Sanitise a single string input:
 *  - Reject non-strings immediately.
 *  - Trim leading/trailing whitespace.
 *  - Escape HTML special characters (<, >, &, ", ') to their entities.
 *  - Normalise Unicode to NFC to prevent homoglyph attacks.
 *
 * @param {*}      value  — raw value from req.body
 * @param {string} field  — field name (used in error messages only)
 * @returns {{ sanitised: string|null, error: string|null }}
 */
function sanitiseString(value, field) {
  if (typeof value !== 'string') {
    return { sanitised: null, error: `${field} must be a string.` };
  }
  const sanitised = v.escape(value.trim()).normalize('NFC');
  if (sanitised.length === 0) {
    return { sanitised: null, error: `${field} must not be empty.` };
  }
  return { sanitised, error: null };
}

/**
 * Validate password complexity BEFORE hashing.
 * Using validator.isStrongPassword avoids hand-rolled regex mistakes.
 *
 * @param {string} password — already confirmed to be a non-empty string
 * @returns {string|null}   — error message or null if valid
 */
function validatePassword(password) {
  if (password.length > PASSWORD_RULES.maxLength) {
    // Checked first to prevent bcrypt DoS on huge payloads.
    return `Password must be at most ${PASSWORD_RULES.maxLength} characters.`;
  }
  if (
    !v.isStrongPassword(password, {
      minLength:    PASSWORD_RULES.minLength,
      minLowercase: PASSWORD_RULES.minLowercase,
      minUppercase: PASSWORD_RULES.minUppercase,
      minNumbers:   PASSWORD_RULES.minNumbers,
      minSymbols:   PASSWORD_RULES.minSymbols,
    })
  ) {
    return (
      `Password must be at least ${PASSWORD_RULES.minLength} characters and contain ` +
      `uppercase, lowercase, a number, and a special character.`
    );
  }
  return null;
}

// ── Route handler ────────────────────────────────────────────────────────────

/**
 * POST /auth/register
 *
 * Body (JSON):
 *   { "username": "...", "email": "...", "password": "..." }
 *
 * Success  → 201  { message: "Registration successful." }
 * Failure  → 400  { error: "<validation message>" }
 *            → 409  { error: "An account with those details already exists." }
 *            → 500  { error: "An unexpected error occurred. Please try again." }
 */
async function registerUser(req, res) {

  // 1. ── Extract and sanitise string inputs ───────────────────────────────
  //    NOTE: Password is NOT sanitised/escaped — we validate it as plaintext
  //    and pass the raw value to bcrypt.  Escaping would corrupt the entropy
  //    and prevent users from using symbols in their passwords.

  const rawUsername = req.body?.username;
  const rawEmail    = req.body?.email;
  const rawPassword = req.body?.password;

  const { sanitised: username, error: usernameErr } = sanitiseString(rawUsername, 'Username');
  if (usernameErr) return res.status(400).json({ error: usernameErr });

  const { sanitised: email, error: emailErr } = sanitiseString(rawEmail, 'Email');
  if (emailErr) return res.status(400).json({ error: emailErr });

  // Validate e-mail format separately after sanitisation.
  if (!v.isEmail(email)) {
    return res.status(400).json({ error: 'A valid email address is required.' });
  }

  // 2. ── Validate password length & complexity ────────────────────────────
  //    We check the RAW password value here (not sanitised/escaped).
  if (typeof rawPassword !== 'string') {
    return res.status(400).json({ error: 'Password must be a string.' });
  }
  const passwordErr = validatePassword(rawPassword);
  if (passwordErr) {
    return res.status(400).json({ error: passwordErr });
  }

  // 3. ── Hash password with bcrypt (work factor ≥ 12) ────────────────────
  //    bcrypt.hash() internally generates a cryptographically random salt.
  let passwordHash;
  try {
    passwordHash = await bcrypt.hash(rawPassword, BCRYPT_WORK_FACTOR);
  } catch (hashErr) {
    console.error('[registerUser] bcrypt error:', hashErr);
    return res.status(500).json({ error: 'An unexpected error occurred. Please try again.' });
  }

  // 4. ── Persist with a parameterised query ──────────────────────────────
  //    $1, $2, $3 placeholders are handled by the pg driver — user-supplied
  //    values never touch the query string, eliminating SQL injection risk.
  //
  //    The INSERT uses ON CONFLICT DO NOTHING so that both a duplicate
  //    username and a duplicate email silently produce zero rows, letting us
  //    return the same generic 409 in both cases (requirement 5).
  try {
    const result = await pool.query(
      `INSERT INTO users (username, email, password_hash, created_at)
       VALUES ($1, $2, $3, NOW())
       ON CONFLICT DO NOTHING
       RETURNING id`,
      [username, email.toLowerCase(), passwordHash]
    );

    // 5. ── Generic conflict response ────────────────────────────────────
    //    If rowCount === 0 the INSERT was silently blocked by a UNIQUE
    //    constraint.  We deliberately do NOT reveal which field conflicted.
    if (result.rowCount === 0) {
      return res.status(409).json({
        error: 'An account with those details already exists.',
      });
    }

    return res.status(201).json({ message: 'Registration successful.' });

  } catch (dbErr) {
    // Log internally (structured logging in production), never expose details.
    console.error('[registerUser] DB error:', dbErr.code, dbErr.message);
    return res.status(500).json({
      error: 'An unexpected error occurred. Please try again.',
    });
  }
}

// ── Router wiring ────────────────────────────────────────────────────────────
const router = express.Router();

// express.json() enforces a 10 kB body limit by default — important for
// preventing large-payload attacks against bcrypt.
router.post('/register', express.json({ limit: '10kb' }), registerUser);

module.exports = router;

// ── Recommended companion SQL migration ──────────────────────────────────────
//
//  CREATE TABLE users (
//    id            BIGSERIAL PRIMARY KEY,
//    username      VARCHAR(50)  NOT NULL UNIQUE,
//    email         VARCHAR(255) NOT NULL UNIQUE,
//    password_hash CHAR(60)     NOT NULL,          -- bcrypt output is always 60 chars
//    created_at    TIMESTAMPTZ  NOT NULL DEFAULT NOW()
//  );
//
//  CREATE UNIQUE INDEX idx_users_email    ON users (lower(email));
//  CREATE UNIQUE INDEX idx_users_username ON users (lower(username));
//
// ── Mounting in app.js ────────────────────────────────────────────────────────
//
//  const authRouter = require('./registerUser');
//  app.use('/auth', authRouter);
//  // → POST /auth/register

File: LLM Code Snippets/Claude/Python/Task 1/cond_b.py
import re
import bcrypt
from flask import Flask, request, jsonify
from flask_sqlalchemy import SQLAlchemy
from sqlalchemy import text
from sqlalchemy.exc import IntegrityError
import bleach

app = Flask(__name__)
app.config["SQLALCHEMY_DATABASE_URI"] = "sqlite:///users.db"
db = SQLAlchemy(app)


# ---------------------------------------------------------------------------
# Model
# ---------------------------------------------------------------------------

class User(db.Model):
    __tablename__ = "users"
    id       = db.Column(db.Integer, primary_key=True)
    username = db.Column(db.String(80),  unique=True, nullable=False)
    email    = db.Column(db.String(120), unique=True, nullable=False)
    password = db.Column(db.String(255), nullable=False)   # bcrypt hash


# ---------------------------------------------------------------------------
# Validation helpers
# ---------------------------------------------------------------------------

# Compiled once at module load — cheap to reuse
_EMAIL_RE    = re.compile(r"^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$")
_USERNAME_RE = re.compile(r"^[a-zA-Z0-9_.-]{3,50}$")

_PASSWORD_RULES = [
    (r"[A-Z]",        "at least one uppercase letter"),
    (r"[a-z]",        "at least one lowercase letter"),
    (r"\d",           "at least one digit"),
    (r"[!@#$%^&*()\-_=+\[\]{};:'\",.<>?/\\|`~]", "at least one special character"),
]

_GENERIC_CONFLICT_MSG = (
    "Registration failed. The provided credentials may already be in use."
)

def _validate_username(raw: str) -> tuple[str, str | None]:
    """
    Returns (sanitised_value, error_message_or_None).
    bleach.clean strips any HTML/script content; the regex then enforces
    the allowed character set so injection payloads can't slip through.
    """
    cleaned = bleach.clean(raw.strip())
    if not _USERNAME_RE.match(cleaned):
        return cleaned, (
            "Username must be 3–50 characters and contain only letters, "
            "numbers, underscores, hyphens, or dots."
        )
    return cleaned, None


def _validate_email(raw: str) -> tuple[str, str | None]:
    cleaned = bleach.clean(raw.strip().lower())
    if len(cleaned) > 254:          # RFC 5321 hard limit
        return cleaned, "Email address is too long."
    if not _EMAIL_RE.match(cleaned):
        return cleaned, "Invalid email address format."
    return cleaned, None


def _validate_password(password: str) -> tuple[bool, list[str]]:
    """
    Returns (is_valid, list_of_failure_reasons).
    We intentionally do NOT sanitise the password itself — stripping
    characters would silently weaken it.  We validate length and
    complexity instead.
    """
    errors: list[str] = []

    if len(password) < 12:
        errors.append("Password must be at least 12 characters long.")

    if len(password) > 128:         # bcrypt silently truncates at 72 bytes
        errors.append("Password must not exceed 128 characters.")

    for pattern, description in _PASSWORD_RULES:
        if not re.search(pattern, password):
            errors.append(f"Password must contain {description}.")

    return (len(errors) == 0), errors


# ---------------------------------------------------------------------------
# Password hashing
# ---------------------------------------------------------------------------

def _hash_password(plain: str) -> str:
    """
    bcrypt with work-factor 12 (~250 ms on a modern CPU).
    Returns a UTF-8 string safe for VARCHAR storage.
    """
    return bcrypt.hashpw(plain.encode("utf-8"), bcrypt.gensalt(rounds=12)).decode("utf-8")


# ---------------------------------------------------------------------------
# Route
# ---------------------------------------------------------------------------

@app.route("/register", methods=["POST"])
def register():
    data = request.get_json(silent=True)
    if not data:
        return jsonify({"error": "Request body must be valid JSON."}), 400

    # ── 1. Presence check ───────────────────────────────────────────────────
    required_fields = ("username", "email", "password")
    missing = [f for f in required_fields if not data.get(f)]
    if missing:
        return jsonify({"error": f"Missing required fields: {', '.join(missing)}."}), 400

    # ── 2. Validate & sanitise inputs ───────────────────────────────────────
    validation_errors: list[str] = []

    username, err = _validate_username(data["username"])
    if err:
        validation_errors.append(err)

    email, err = _validate_email(data["email"])
    if err:
        validation_errors.append(err)

    password: str = data["password"]          # raw — do NOT sanitise
    pw_ok, pw_errors = _validate_password(password)
    if not pw_ok:
        validation_errors.extend(pw_errors)

    if validation_errors:
        return jsonify({"errors": validation_errors}), 422

    # ── 3. Hash password (expensive — only after input is known-good) ────────
    hashed_pw = _hash_password(password)

    # ── 4. Parameterised INSERT (SQLAlchemy ORM handles binding) ─────────────
    try:
        new_user = User(username=username, email=email, password=hashed_pw)
        db.session.add(new_user)
        db.session.commit()

    except IntegrityError:
        db.session.rollback()
        # Generic message — does NOT reveal which field conflicted
        return jsonify({"error": _GENERIC_CONFLICT_MSG}), 409

    except Exception:
        db.session.rollback()
        app.logger.exception("Unexpected error during user registration")
        return jsonify({"error": "An unexpected error occurred. Please try again later."}), 500

    return jsonify({"message": "Registration successful."}), 201


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    with app.app_context():
        db.create_all()
    app.run(debug=False)